common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -37,56 +38,88 @@ import (
 	"time"
 
 	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
 )
 
 var _ CertProvider = (*localStoreCertProvider)(nil)
 var _ CertExpirationChecker = (*localStoreCertProvider)(nil)
 
-type localStoreCertProvider struct {
-	sync.RWMutex
-
-	tlsSettings       *config.GroupTLS
-	workerTLSSettings *config.WorkerTLS
-
+type certCache struct {
 	serverCert          *tls.Certificate
 	clientCert          *tls.Certificate
+	workerCert          *tls.Certificate
 	clientCAPool        *x509.CertPool
 	serverCAPool        *x509.CertPool
 	serverCAsWorkerPool *x509.CertPool
 	clientCACerts       []*x509.Certificate // copies of certs in the clientCAPool CertPool for expiration checks
 	serverCACerts       []*x509.Certificate // copies of certs in the serverCAPool CertPool for expiration checks
 	serverCACertsWorker []*x509.Certificate // copies of certs in the serverCAsWorkerPool CertPool for expiration checks
+}
+
+type localStoreCertProvider struct {
+	sync.RWMutex
 
+	tlsSettings          *config.GroupTLS
+	workerTLSSettings    *config.WorkerTLS
 	isLegacyWorkerConfig bool
 	legacyWorkerSettings *config.ClientTLS
+
+	certs *certCache
+
+	ticker *time.Ticker
+	stop   chan bool
+	logger log.Logger
 }
 
 type loadOrDecodeDataFunc func(item string) ([]byte, error)
 
-type x509CertFetcher func() (*x509.Certificate, error)
-type x509CertPoolFetcher func() (*x509.CertPool, error)
 type tlsCertFetcher func() (*tls.Certificate, error)
 
+func (s *localStoreCertProvider) Initialize(refreshInterval time.Duration) {
+
+	s.logger = log.NewDefaultLogger()
+
+	if refreshInterval != 0 {
+		s.stop = make(chan bool)
+		s.ticker = time.NewTicker(refreshInterval)
+		go s.refreshCerts()
+	}
+}
+
+func (s *localStoreCertProvider) Close() {
+
+	if s.ticker != nil {
+		s.ticker.Stop()
+	}
+	if s.stop != nil {
+		s.stop <- true
+		close(s.stop)
+	}
+}
+
 func (s *localStoreCertProvider) FetchServerCertificate() (*tls.Certificate, error) {
 
 	if s.tlsSettings == nil {
 		return nil, nil
 	}
-	return s.fetchCertificate(&s.serverCert, s.tlsSettings.Server.CertFile, s.tlsSettings.Server.CertData,
-		s.tlsSettings.Server.KeyFile, s.tlsSettings.Server.KeyData)
+	certs, err := s.getCerts()
+	if err != nil {
+		return nil, err
+	}
+	return certs.serverCert, nil
 }
 
 func (s *localStoreCertProvider) FetchClientCAs() (*x509.CertPool, error) {
```

Comment:
it is quite possible that FetchClientCAs reads one version and then a subsequent FetchServerCertificate fetches a different version of the data (because it was refreshed inbetween). Are there any scenarios where the caller will get stuck if it is doing multiple API calls to the provider and it happens to get caught between reloads? Or will this just be a transient issue that resolves itself?

Reply:
I believe this is possible in any case. Shouldn't be a problem though because these calls are made on every connection attempt. So, worst case it should be a transient error.